### PR TITLE
Minor change to allow config to be read from a custom json file

### PR DESCRIPTION
**Pull request by QANTau**
I would like to be able to read from a custom json file located in a path external to the project directory.  The intention is that the custom file could contain data that I never want to have in a git managed directory (usernames, passwords, etc).

### CODE REVIEW

**Comment by jaredbroad** [comment]
Thank you for PR @QANTau! We really appreciate community contributions! 

IMHO the right way to do this would be to add the config file to your git ignore to keep it out of the commits, or change the const config file name to a path outside of the git directory.

Is the submitted solution is for having two config files? (it is a method level argument which infers it would it change from request to request). In general this would increase entropy and might later cause bugs.

Also its slightly counter intuitive / non-deterministic:
```
Config.Get("my-password", "passwords.json");
Config.Get("algorithm"); //Not Found; because custom config file overrides the default config
```

I think it might be time we convert this to use the .NET app.config XML's since that is the "right" way to do configuration in .NET. If you're interested you could start this by making the Config implement IConfigurationHandler like @bdilber's ILogHandler abstracts the logging - then the current configuration class would be a "JsonConfigHandler" - https://github.com/QuantConnect/Lean/blob/master/Logging/Log.cs#L44

**Comment by QANTau** [comment]
The intention was to be able to have a single configuration file that could be used for multiple purposes when running on a local instance (although no reason why this couldn't be used to work with multiple different config files when running in the cloud).  In my current tests, I have my QC login details (for Rest API Access) and broker credentials set in a central file so I don't need to have this duplicated across multiple projects.

I can adjust the code to ensure that the default config.json file name is reset after each 'custom file' read to resolve one of the issues (ie. by settings the _loaded flag to false after a 'custom file' read).

For me (personally), I think that the current mechanism of using json files is simple and I'm not sure if moving to XML adds much value.

Another option may be to create a CustomConfig class based on the current Config class to read from a non-standard file.

**Comment by mchandschuh** [comment]
It seems the impetus for this change is to protect credentials from being
put into the git repository. It sounds like the easiest way to do this is
to just locally ignore that file using the gitignore. Is there a reason
this solution doesn't work?

Having multiple config files adds complexity to something that in my
opinion doesn't really need it.

As for needing to maintain multiple copies of the configuration, maybe we
can set those files up as links so only one file exists but multiple
projects reference it.

File: Configuration/Config.cs [hunk 1]
```diff
@@ -83,13 +83,27 @@ private static void Initialize()
         }
         
         /// <summary>
+        /// Sets a custom file name
+        /// </summary>
+        /// <param name="customFileName"></param>
+        private static void SetCustomFileName(string customFileName)
+        {
+            _loaded = false;
```
Comment (mchandschuh) [inline Configuration/Config.cs]: This is not thread-safe at all. Imagine the case where thread 1 comes in and requests the normal config, but is then suspended, thread 2 comes in for a custom file and sets the _config static field and is then suspended. Thread 1 resumes and reads value from the new value that thread 2 set.

**Comment by mchandschuh** [comment]
Another option, why not use the regular configuration file to add an item that points to your configuration file. It seems this is 100% custom code that needs these values, so maybe an out of band solution is more appropriate.

**Comment by QANTau** [comment]
No problems.  Agreed that this not thread-safe.  I currently have an entry in config.json that references an external file and was looking to use the existing methods to read the file.  I'll remove the pull request and just implement external config reader code.

**Pull request closed without merging**